director/sort.go

```diff
@@ -165,29 +165,59 @@ func getLatLong(addr netip.Addr) (lat float64, long float64, err error) {
 	return
 }
 
+func getClientLatLong(addr netip.Addr) (coord Coordinate, ok bool) {
+	var err error
+	coord.Lat, coord.Long, err = getLatLong(addr)
+	ok = (err == nil && !(coord.Lat == 0 && coord.Long == 0))
+	if err != nil {
+		log.Warningf("failed to resolve lat/long for address %s: %v", addr, err)
+	}
+	return
+}
+
 // Sort serverAds based on the IP address of the client with shorter distance between
 // server IP and client having higher priority
 func sortServerAdsByIP(addr netip.Addr, ads []server_structs.ServerAd) ([]server_structs.ServerAd, error) {
-	distances := make(SwapMaps, len(ads))
-	lat, long, err := getLatLong(addr)
-	// If we don't get a valid coordinate set for the incoming address, either because
-	// of an error or the null address, we randomize the output
-	isInvalid := (err != nil || (lat == 0 && long == 0))
+	// Each entry in weights will map a priority to an index in the original ads slice.
+	// A larger weight is a higher priority.
+	weights := make(SwapMaps, len(ads))
+	sortMethod := param.Director_CacheSortMethod.GetString()
+
+	// For each ad, we apply the configured sort method to determine a priority weight.
 	for idx, ad := range ads {
-		if isInvalid || (ad.Latitude == 0 && ad.Longitude == 0) {
-			// Unable to compute distances for this server; just do random distances.
-			// Note that valid distances are between 0 and 1, hence (1 + random) is always
-			// going to be sorted after valid distances.
-			distances[idx] = SwapMap{1 + rand.Float64(), idx}
-		} else {
-			distances[idx] = SwapMap{distanceOnSphere(lat, long, ad.Latitude, ad.Longitude),
-				idx}
+		switch sortMethod {
+		case "distance":
+			clientCoord, ok := getClientLatLong(addr)
+			if !ok {
+				// Unable to compute distances for this server; just do random distances.
+				// Below we sort weights in descending order, so we assign negative value here,
+				// causing them to always be at the end of the sorted list.
+				weights[idx] = SwapMap{0 - rand.Float64(), idx}
+			} else {
+				weights[idx] = SwapMap{distanceWeight(clientCoord, ad),
+					idx}
+			}
+		case "distanceAndLoad":
+			clientCoord, ok := getClientLatLong(addr)
+			if !ok {
+				weights[idx] = SwapMap{0 - rand.Float64(), idx}
+			} else {
+				// Each server ad will have a load value that we can use for sorting
+				weights[idx] = SwapMap{distanceAndLoadWeight(clientCoord, ad),
+					idx}
+			}
+		case "random":
+			weights[idx] = SwapMap{rand.Float64(), idx}
+		default:
+			return nil, errors.Errorf("Invalid sort method '%s' set in Director.CacheSortMethod", param.Director_CacheSortMethod.GetString())
```

Comment:
```suggestion
			return nil, errors.Errorf("invalid sort method '%s' set in Director.CacheSortMethod", param.Director_CacheSortMethod.GetString())
```
Do we want to mention available options in the error message as well?